### DISCUSSION TITLE
Best way to do Semi-supervised CL

### DISCUSSION THREAD

**Discussion by Mattdl** [ContinualAI/avalanche#663] [Q&A] [answered]
I'm trying out some semi-supervised settings, but I'm wondering what would be the best way to implement this in Avalanche?
Say we have the typical CIFAR10 split into 5 tasks fo 2 classes, how do I make a split for the small labeled subset vs the large unlabeled subset per experience?
Thanks!

**Comment by AntonioCarta** [comment, marked as answer]
For the benchmark definition, maybe the easiest thing to do is to have two parallel streams, one is labelled, the other unlabeled. The strategy will use both in parallel.

**Reply by Mattdl** [reply]
Seems like a good solution! Implementation-wise, would it be similar to the benchmark generator: `benchmark_with_validation_stream` ?
In fact, I think I could just use the `benchmark_with_validation_stream` and ignore the labels of the 'validation' (unlabeled) stream!
Where does the iteration over the streams happen exactly? Can I process two minibatches, one from each stream, and then make an update?

**Reply by AntonioCarta** [reply]
you can use one of the generic generators, like `create_multi_dataset_generic_benchmark`.

> Where does the iteration over the streams happen exactly? Can I process two minibatches, one from each stream, and then make an update?

You need a custom strategy for that. I don't know how your strategy will process the data so it's difficult for me to guess what you need.

I imagine you will have a `train(sup_exp, unsup_exp)` with the additional `unsup_exp` argument, and then a custom dataloader/forward/criterion to iterate both in parallel (if you need to do it).
